src/editor.rs

```diff
@@ -408,18 +410,32 @@ impl Editor {
     }
 
     fn duplicate_current_row(&mut self) {
+        self.copy_current_row(true);
+        self.paste_current_row();
+    }
+
+    fn copy_current_row(&mut self, is_preserving: bool) {
         if let Some(row) = self.current_row() {
-            let new_row = Row::new(row.chars.clone());
-            self.n_bytes += new_row.chars.len() as u64;
-            self.rows.insert(self.cursor.y + 1, new_row);
-            self.update_row(self.cursor.y + 1, false);
-            self.cursor.y += 1;
-            self.dirty = true;
-            // The line number has changed
-            self.update_screen_cols();
+            self.copied_row = row.chars.clone();
+        }
+
+        if !is_preserving {
+            self.delete_current_row();
         }
     }
 
+    fn paste_current_row(&mut self) {
+        if self.copied_row.is_empty() {
+            return;
+        }
+        self.n_bytes += self.copied_row.len() as u64;
+        self.rows.insert(self.cursor.y + 1, Row::new(self.copied_row.clone()));
```

Comment:
This causes a panic if the cursor is at the bottom of the editor (at a line that doesn't exist yet)

Reply:
Fixed this issue